### PR TITLE
feat: add network failure detection and Azure network error handling (ARO-24301)

### PR DESCRIPTION
**Pull request by RadekCap**
## Description

Add structured network failure detection and Azure network error handling to improve diagnostics
when network-related failures occur during cluster deployment and monitoring.

JIRA: https://redhat.atlassian.net/browse/ARO-24301

## Changes Made

- Add `DetectNetworkError()` with 6 error categories: DNS resolution, connection refused, connection timeout, TLS/certificate, connection reset, API server errors — each with remediation steps
- Add `FormatNetworkError()` for human-readable diagnostic output
- Extend `DetectAzureError()` with 5 Azure network error types: DNS zone, NSG, load balancer, VNet, subnet errors
- Enhance `MonitorClusterUntilReady`, `MonitorControlPlaneUntilReady`, `MonitorNodesUntilAvailable`, and `MonitorClusterUntilDeleted` to check management cluster health when monitoring fails — abort after 3 consecutive unreachable failures with structured guidance
- Enhance `ApplyWithRetryInNamespace` to log network error context (category + remediation) alongside retry messages
- Add 57 test cases: `TestDetectNetworkError` (27 cases), `TestFormatNetworkError` (3 cases), extend `TestDetectAzureError` (13 new Azure network error cases), plus nil/edge cases
- Update CLAUDE.md helper listing

## Configuration Changes

No new environment variables.

## Additional Notes

The existing `isRetryableKubectlError()` decides *whether* to retry; the new `DetectNetworkError()` explains *what went wrong* and *how to fix it*. They complement each other — `isRetryableKubectlError` is unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **Tests**
  * Added unit coverage for structured network diagnostics, including classification and formatted “Diagnostic steps” output.
  * Expanded Azure networking failure detection tests (DNS zones, NSG/security rules, Load Balancers, VNet conflicts, subnet issues).

* **Bug Fixes**
  * Improved monitoring reliability for cluster/control plane/node readiness by tolerating transient failures with a consecutive-failure threshold.
  * Added management-cluster reachability checks to distinguish network issues and abort after repeated unreachable conditions; applied the same logic during cluster deletion.

### CODE REVIEW

File: test/cluster_monitor.go [hunk 1]
```diff
@@ -119,6 +119,31 @@ type ConditionsSummary struct {
...
+const maxConsecutiveMonitorFailures = 3
+
+// checkManagementClusterHealth runs a lightweight kubectl command against the management cluster
+// to determine if the cluster API server is reachable. Returns nil if healthy, or a descriptive
+// error with remediation guidance from DetectNetworkError if unreachable.
+func checkManagementClusterHealth(t *testing.T, kubeContext string) error {
+	t.Helper()
+
+	// #nosec G204 -- kubeContext is validated upstream as RFC 1123 compliant
+	cmd := exec.Command("kubectl", "--context", kubeContext, "get", "ns", "--request-timeout=5s")
```
[the remark this answers is not in the document]
Comment (RadekCap) [reply test/cluster_monitor.go:133]: **Not Implementing** - Finding #1

**Rationale**: The improvement doesn't justify the code churn.

**Reasoning**:
- The 5s `--request-timeout` kubectl flag already provides sufficient timeout protection
- Adding `context.Context` requires changing the function signature and threading context through all 4 monitoring loop callers
- CodeRabbit itself marks this as "Nitpick / Trivial / Poor tradeoff" and notes it "may be deferred"
- No functional or safety benefit over the current implementation

File: test/cluster_monitor.go [hunk 2]
```diff
@@ -119,6 +119,31 @@ type ConditionsSummary struct {
...
+	if err == nil {
+		return nil
+	}
+
+	combined := string(output) + " " + err.Error()
+	netErr := DetectNetworkError(combined)
+	if netErr != nil {
+		return fmt.Errorf("management cluster unreachable (%s): %s", netErr.ErrorType, netErr.Message)
+	}
+	return fmt.Errorf("management cluster health check failed: %v", err)
```
[the remark this answers is not in the document]
Comment (RadekCap) [reply test/cluster_monitor.go]: **Implemented** - Finding #2

**Change Made**: Changed `%v` to `%w` for proper Go 1.13+ error wrapping, enabling `errors.Is()` and `errors.Unwrap()` on the returned error.

**Commit**: `5808dd4`

**Comment by RadekCap** [comment]
**Pre-merge Check - False Positive**: Docstring Coverage

**Rationale**: All new exported types and functions have docstrings:
- `NetworkErrorInfo` struct — has docstring
- `DetectNetworkError` function — has multi-line docstring
- `FormatNetworkError` function — has docstring
- `checkManagementClusterHealth` function — has multi-line docstring
- `maxConsecutiveMonitorFailures` constant — has docstring

**Details**:
- The 75% coverage likely includes struct field-level comments, which are not present — following the exact same pattern as the existing `AzureErrorInfo` struct in the same file
- Go convention uses inline comments for struct fields, not docstrings
- Adding field comments would be inconsistent with the codebase pattern

**Comment by RadekCap** [comment]
## AI Review Pipeline Summary

| Category | Details |
|----------|---------|
| **Self-Review** | 5 findings found, 1 accepted, 4 denied |
| **Security Review** | 0 findings |
| **CodeRabbit** | 1 round, 1 accepted, 1 denied |
| **Pre-merge Checks** | 1 failure (false positive) |
| **Qodo** | 0 findings |

### Accepted Findings

| # | Source | Description | File | Commit |
|---|--------|-------------|------|--------|
| 1 | Self-Review | Add failure-context guard to `backendaddresspool` detection | `test/helpers.go` | `80f7dba` |
| 2 | CodeRabbit | Use `%w` for proper error wrapping (Go 1.13+) | `test/cluster_monitor.go:144` | `5808dd4` |

### Denied Findings

| # | Source | Description | Rationale |
|---|--------|-------------|-----------|
| 1 | Self-Review | VNet `"address" && "overlap"` too broad | Only called on error output, combination is specific enough |
| 2 | Self-Review | `"certificate"` TLS check too broad | Only called on error output where "certificate" indicates TLS issue |
| 3 | Self-Review | `exec.Command` instead of `RunCommandQuiet` | Intentional — avoids log spam during transient health check failures |
| 4 | Self-Review | Duplicated consecutive-failure logic | Per CLAUDE.md: "Three similar lines is better than a premature abstraction" |
| 5 | CodeRabbit | Use `exec.CommandContext` for cancellation | CodeRabbit marks as "Nitpick / Poor tradeoff"; 5s kubectl timeout is sufficient |

---
All CodeRabbit threads resolved: Yes. Stale review dismissed: Yes.

🤖 Generated by `/ai-review` pipeline

File: test/cluster_monitor.go [hunk 3]
```diff
@@ -119,6 +119,48 @@ type ConditionsSummary struct {
...
+func handleMonitorFailure(t *testing.T, kubeContext string, iteration int, consecutiveFailures *int, monitorErr error) error {
+	t.Helper()
+	*consecutiveFailures++
+	t.Logf("[%d] Warning: failed to get cluster status: %v", iteration, monitorErr)
+
+	if healthErr := checkManagementClusterHealth(t, kubeContext); healthErr != nil {
+		t.Logf("[%d] ⚠️  %v", iteration, healthErr)
+		if *consecutiveFailures >= maxConsecutiveMonitorFailures {
+			return fmt.Errorf("aborting after %d consecutive monitoring failures: %w", *consecutiveFailures, healthErr)
+		}
```
[the remark this answers is not in the document]
Comment (RadekCap) [reply test/cluster_monitor.go:159]: **Implemented** - Moved consecutiveFailures increment inside the healthErr block so the counter only tracks management-cluster-unreachable failures. Added reset to 0 when management cluster is healthy. Commit: 379a29c

File: test/helpers.go [hunk 4]
```diff
@@ -2236,9 +2254,9 @@ func DetectAzureError(output string) *AzureErrorInfo {
 		}
 	}
 
-	// Resource group not found
+	// Resource group not found (exclude DNS zone messages that also mention resource group)
 	if strings.Contains(lowerOutput, "resourcegroupnotfound") ||
-		strings.Contains(lowerOutput, "resource group") && strings.Contains(lowerOutput, "not found") {
+		(strings.Contains(lowerOutput, "resource group") && strings.Contains(lowerOutput, "not found") && !strings.Contains(lowerOutput, "dns zone")) {
```
[the remark this answers is not in the document]
Comment (RadekCap) [reply test/helpers.go]: **Implemented** - Restructured the condition so the DNS zone exclusion applies to the entire resource group check, preventing resourcegroupnotfound from short-circuiting past the dns zone guard. Commit: d4a2cb5

**Comment by RadekCap** [comment]
**Pre-merge Check — False Positive**: Docstring Coverage

**Rationale**: All 5 new functions/types added in this PR (`NetworkErrorInfo`, `DetectNetworkError`, `FormatNetworkError`, `checkManagementClusterHealth`, `handleMonitorFailure`) have docstrings. The 76.47% coverage is a pre-existing gap in other functions, not introduced by this PR.

**Details**:
- All new exported symbols have proper Go doc comments
- All new unexported functions have doc comments explaining their purpose
- The coverage gap comes from existing functions in `helpers.go` and `cluster_monitor.go` that predate this PR

**Comment by RadekCap** [comment]
## AI Review Pipeline Summary

| Category | Details |
|----------|---------|
| **Self-Review** | 0 new findings (already addressed in prior commit) |
| **Security Review** | Skipped (self-review was clean) |
| **CodeRabbit** | 1 round, 2 accepted, 0 denied |
| **Pre-merge Checks** | 1 failure (false positive — docstring coverage) |
| **Qodo** | No findings |

### Accepted Findings

| # | Source | Description | File | Commit |
|---|--------|-------------|------|--------|
| 1 | CodeRabbit | Fix circuit-breaker counter semantics — only increment when management cluster is also unreachable | `test/cluster_monitor.go:152` | `379a29c` |
| 2 | CodeRabbit | Fix DNS zone exclusion short-circuit — apply exclusion to entire resource group condition | `test/helpers.go:2258` | `d4a2cb5` |

---
All CodeRabbit threads resolved: Yes. Stale review dismissed: Yes. PR description updated: Already current (no changes made).

🤖 Generated by `/ai-review` pipeline

**Review by marek-veber — APPROVED** [review]
/lgtm
